Fix K0smotronControlPlane version mismatch for Cluster API compatibility

kahirokunn opened the pull request:
## Main Content

This PR resolves the K0smotronControlPlane and Cluster API compatibility issue reported in Issue #1052:

### Problem
- Version format discrepancy between `spec.version` (e.g., `v1.33.1`) and `status.version` (e.g., `v1.33.1-k0s.0`)
- Cluster API's MachineSet Preflight Checks incorrectly recognize this as "upgrading in progress"
- MachineDeployment scale-up operations are indefinitely suspended

### Solution
1. **Version normalization function**: Unified version format using the `FormatStatusVersion` function
2. **Complete version information retention**: Store complete k0s version information in the `K0sVersion` field

The detailed PR description is saved in the `pr-title-and-description.md` file. This PR enables k0smotron to function correctly as a standard control plane provider within the Cluster API ecosystem.

jnummelin commented on internal/controller/controlplane/k0smotron_controlplane_controller.go:
```diff
@@ -344,6 +344,23 @@ func (c *K0smotronController) ensureCertificates(ctx context.Context, cluster *c
+func FormatStatusVersion(specVersion, statusVersion string) string {
+	specHasK0sSuffix := strings.Contains(specVersion, "-k0s.")
+
+	// spec.versionの形式に合わせてstatus.versionを調整
```
Japanese text? 🧐

jnummelin commented on internal/controller/controlplane/k0smotron_controlplane_controller.go:
```diff
@@ -344,6 +344,23 @@ func (c *K0smotronController) ensureCertificates(ctx context.Context, cluster *c
+
+	// spec.versionの形式に合わせてstatus.versionを調整
+	if !specHasK0sSuffix && strings.Contains(statusVersion, "-k0s.") {
+		// spec.versionに-k0s.サフィックスがない場合は、status.versionからも除去
```
Japanese again, pls change to English 🙂

kahirokunn commented:
I've added e2e tests that cover the MachineDeployment scaling scenario. The tests verify:
Version format consistency between spec.version and status.version
MachineDeployment scale-up operations work correctly (no longer hang due to version mismatch)
Both status.version and status.k0sVersion fields are properly populated
Thanks for the feedback!

kahirokunn commented:
Rebased with the latest main

makhov commented on inttest/capi-docker-machinedeployment/capi_docker_test.go:
```diff
@@ -150,8 +169,88 @@ func (s *CAPIDockerSuite) TestCAPIDocker() {
+}
+
+// verifyK0smotronControlPlaneVersionFormat verifies that the version format is consistent
+func (s *CAPIDockerSuite) verifyK0smotronControlPlaneVersionFormat(clusterName, expectedSpecVersion string, expectK0sSuffix bool) error {
```
I think it's too complex func for the task. Could we just keep it simple?

kahirokunn replied:
Thx. I did it.

makhov commented on inttest/capi-docker-machinedeployment/capi_docker_test.go:
```diff
@@ -150,8 +169,88 @@ func (s *CAPIDockerSuite) TestCAPIDocker() {
+		if !strings.Contains(status.Version, "-k0s.") {
+			return fmt.Errorf("expected status.version to contain '-k0s.' suffix, but got: %s", status.Version)
+		}
+		if status.Version != expectedSpecVersion {
```
We use testify library for asserts

kahirokunn replied:
Thx. I did it.

makhov reviewed (changes requested):
Thanks for the PR! I appreciate the effort, but I think a lot of changes here are out of the scope, I think we can simplify it a lot

kahirokunn commented on inttest/capi-docker-machinedeployment/CHANGES.md:
```diff
@@ -0,0 +1,72 @@
+# Changes Summary for K0smotronControlPlane Version Format E2E Test
```
Thx. I deleted it.

kahirokunn commented on inttest/capi-docker-machinedeployment/README.md:
```diff
@@ -0,0 +1,57 @@
+# CAPI Docker MachineDeployment Version Format Test
```
Thx. I deleted it.

kahirokunn commented on inttest/capi-docker-clusterclass/capi_docker_clusterclass_test.go:
```diff
@@ -63,6 +63,121 @@ func (s *CAPIDockerClusterClassSuite) SetupSuite() {
+	s.FootlooseSuite.TearDownSuite()
+}
+
+func (s *CAPIDockerClusterClassSuite) dumpClusterState() {
```
Sorry, this commit is a remnant from when I was trying to fix a flaky CI issue on this branch.
I moved it to PR #1073, and some issues have been resolved, but the root cause has not been fixed.

kahirokunn commented on inttest/basic/basic_test.go:
```diff
@@ -65,6 +65,9 @@ func (s *BasicSuite) TestK0sGetsUp() {
 	s.Require().NoError(common.WaitForStatefulSet(s.Context(), kc, "kmc-kmc-test", "kmc-test"))
 
+	s.T().Log("waiting for all deployments in kmc-test namespace to be ready")
+	s.Require().NoError(util.WaitForDeploymentsInNamespace(s.Context(), kc, "kmc-test"))
```
Sorry, this commit is a remnant from when I was trying to fix a flaky CI issue on this branch.
I moved it to PR https://github.com/k0sproject/k0smotron/pull/1073, and some issues have been resolved, but the root cause has not been fixed.

kahirokunn commented:
Thank you for the feedback. I apologize for any confusion—because I work in the JST time zone, I assumed the team was still offline and briefly tested some CI fixes on this branch. I’ve now addressed all your comments and reverted the temporary changes. When you have a moment, could you please review it again? I’d really appreciate it.

kahirokunn commented:
Overcoming flaky tests and achieving a miracle I am thrilled that all CI tests passed. 😍 👍

kahirokunn commented:
Bump up 🤚

kahirokunn commented:
Conflict resolved

cwrau reviewed (approved):
Logic looks good to me, especially since https://github.com/k0sproject/k0smotron/pull/1053/files#diff-0f7bdf411afd541628aba6e0c9feee81f9c85ad4c462de03021792f6ae2bbdc6R116 should catch anything

kahirokunn commented:
I was relieved to hear from you.
Thank you for always working on such wonderful projects!
It's totally fine.
Thank you 🙏

apedriza commented on internal/controller/controlplane/k0smotron_controlplane_controller.go:
```diff
@@ -468,7 +485,11 @@ func (c *K0smotronController) computeStatus(ctx context.Context, cluster types.N
 	kcp.Status.UnavailableReplicas = int32(unavailableReplicas)
 
 	if kcp.Status.ReadyReplicas > 0 {
-		kcp.Status.Version = minimumVersion.String()
+		statusVersion := minimumVersion.String()
+		// Store the full k0s version
+		kcp.Status.K0sVersion = statusVersion
+		// Store the formatted version for Cluster API compatibility
+		kcp.Status.Version = FormatStatusVersion(kcp.Spec.Version, statusVersion)
```
I would store in `.status.k0sVersion` just the [k0s version](https://docs.k0sproject.io/v1.31.3+k0s.0/releases/#:~:text=(k0s.0)%20reflects%20the%20k0s%20version%2C%20which%20is%20built%20on%20top%20of%20the%20certain%20Kubernetes%20version.) and in `.status.version` the k8s version. For example, for `v1.30.0-k0s.0` :

- `.status.k0sVersion` = `k0s.0`
- `.status.version` = `v1.30.0`

this way we avoid redundancy.

kahirokunn commented on internal/controller/controlplane/version_format_test.go:
```diff
@@ -0,0 +1,71 @@
+/*
```
Thank you. I fixed it.

kahirokunn commented on inttest/capi-docker-machinedeployment-no-suffix/capi_docker_test.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+func (s *CAPIDockerSuite) TestCAPIDocker() {
+	s.testCAPIDockerWithVersion("docker-md-test-no-suffix", "v1.27.2", "../../config/samples/capi/docker/cluster-with-machinedeployment-no-suffix.yaml", false)
```
Thx. I fixed it.

kahirokunn commented:
@apedriza I made the corrections as instructed. How does it look? 👀

apedriza reviewed (commented):
from e2e output: `testing: warning: no tests to run`

seems something is missing and test is not running

apedriza commented on e2e/machinedeployment_test.go:
```diff
@@ -0,0 +1,233 @@
+
+func TestMachineDeployment(t *testing.T) {
+	setupAndRun(t, func(t *testing.T) {
+		t.Run("without-k0s-suffix", func(t *testing.T) {
+			machineDeploymentSpec(t, "v1.27.2", "v1.27.2")
+		})
+
+		t.Run("with-k0s-suffix", func(t *testing.T) {
+			machineDeploymentSpec(t, "v1.27.2+k0s.0", "v1.27.2+k0s.0")
+		})
```
I would suggest running only an e2e spec using the classic version format, without involving the k0s version. I recommend avoiding parameterizing the spec function. Everything required for the test should either be self-contained or rely on environment variables.

There is no need to introduce configurability to the spec if only a single variant is intended to run. The unit tests you created cover the different configurations and scenarios

kahirokunn commented on e2e/machinedeployment_test.go:
```diff
@@ -0,0 +1,233 @@
+	err = util.WaitForControlPlaneToBeReady(ctx, bootstrapClusterProxy.GetClient(), controlPlane, util.GetInterval(e2eConfig, testName, "wait-control-plane"))
+	require.NoError(t, err)
+
+	t.Log("Verifying K0smotronControlPlane version format")
```
Thx. I fixed it.

kahirokunn commented:
Conflict resolved

kahirokunn commented:
Will this be available in `v1.7.0` ? 👀

apedriza commented:
@kahirokunn  @broboa have you try to set same version format in `Cluster.spec.version` than `K0smotronControlPlaneTemplate.spec.version` ? for example, `v1.27.2-k0s.0`. MachineDeployment version comes from `Cluster.spec.version` so in that case the comparison succeed. I don't know further implications but should work. Letm me know if works for your case.

broboa commented:
@apedriza are you sure it is exactly `cluster.spec.version` and not `cluster.spec.topology.version`? It seems that `cluster.spec.version` doesnt exist.

broboa commented:
@apedriza defining `cluster.spec.topology` requires both `version` and `class`. A ClusterClass seems like a bin thing to define, or? https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass

kahirokunn commented:
@apedriza 

Thank you for your feedback. I'd like to clarify my understanding and align on the path forward.

## My Initial Understanding

I initially understood this PR was addressing the **presence/absence of the suffix** issue:
- `v1.31.1` (shortened form without suffix)
- `v1.31.1-k0s.0` (full form with suffix)

This mismatch was causing the Cluster API compatibility issue where MachineDeployment scaling would hang.

## Your Actual Proposal (if I understand correctly)

After re-reading your comment, I now understand you're proposing to address a different aspect - the **symbol standardization** between:
- `v1.31.1-k0s.0` (hyphen format for Docker Hub)
- `v1.31.1+k0s.0` (plus format as Kubernetes standard)

Your suggestion is to:
1. Accept only the `+` format from users (e.g., `v1.31.1+k0s.0`)
2. Automatically convert to `-` format internally when creating StatefulSets for Docker Hub
3. Make this Docker Hub technicality transparent to users

I wasn't aware that container tags don't allow `+` symbols. I found that other projects like Helm also auto-convert `+` to `-` (https://github.com/opencontainers/distribution-spec/issues/154#issuecomment-2978081486), so I agree with this automatic conversion approach.

## Moving Forward

While the `+` to `-` conversion is indeed a valuable improvement, this PR currently solves a concrete problem: **the version format mismatch that prevents MachineDeployment from scaling**. This is affecting users today, as reported in #1052 and #1027.

Is there anything specific that needs to be addressed in this PR to move it forward? I'm happy to make any necessary adjustments.

apedriza commented:
> this PR currently solves a concrete problem: **the version format mismatch that prevents MachineDeployment from scaling**. This is affecting users today, as reported in #1052 and #1027.

@kahirokunn have you try to use `<k8s-version>-<k0s-version>` format for the `K0smotronControlPlane.spec.version` ? that should avoid the failing comparison between `.spec.version` and `.status.version` and make the preflight check succeed. Probably is failing is your case because you are using a plain k8s version format for the K0smotronControlPlane. Can you confirm?

kahirokunn commented:
@apedriza Thank you very much for your kind response! 🙏 
However, I am not seeking a temporary workaround. Since k0smotron is a ClusterAPIProvider, I created this PR with the intention of addressing the parts that do not comply with the ClusterAPI specification, making fundamental fixes, and making many people happy. 😆

kahirokunn commented:
Many of the current issues are due to non-compliance with ClusterAPI KEPs.
Compliance with ClusterAPI KEPs will resolve many issues and enable integration with more providers.
And many people will be happy. 👍

jnummelin commented:
@apedriza @kahirokunn I agree with both of you 😂  Let me explain what I'm thinking.

I think we should have some mutating webhook to standardize the version format we take in, like @apedriza is mentioning. This would mean we actually store the version always in the same format "internally".

But even if we'll create something like that say in some coming 1.X.0 version, it still leaves this problem for already existing resources. And for that, we'd need what this PR implements, otherwise those existing cluster will continue to suffer from this since the webhooks only kick-in when users create/modify the resources in the API.

What I'm not sure is the introduction of `K0sVersion` field. As far as I see we don't _really_ use the field for anything. I mean we don't use it internally in k0smotron code for anything nor do we show it to users, other that if they dig deep into the status field. So is this really needed? 🤔

apedriza commented:
> @apedriza @kahirokunn I agree with both of you 😂 Let me explain what I'm thinking.
> 
> I think we should have some mutating webhook to standardize the version format we take in, like @apedriza is mentioning. This would mean we actually store the version always in the same format "internally".
> 
> But even if we'll create something like that say in some coming 1.X.0 version, it still leaves this problem for already existing resources. And for that, we'd need what this PR implements, otherwise those existing cluster will continue to suffer from this since the webhooks only kick-in when users create/modify the resources in the API.
> 
> What I'm not sure is the introduction of `K0sVersion` field. As far as I see we don't _really_ use the field for anything. I mean we don't use it internally in k0smotron code for anything nor do we show it to users, other that if they dig deep into the status field. So is this really needed? 🤔

yes, we definitely need some kind of transformation for the existing resources as proposed in the PR 💯 . For creating new resources, mutating them if they comes with a  plain k8s versions by adding the k0s suffix should also fix the current bug. On the other hand, I think we should write some warning in the webhook about it would be default to `k0s.0` version

kahirokunn commented on inttest/capi-docker-clusterclass-k0smotron/capi_docker_clusterclass_k0smotron_test.go:
```diff
@@ -184,7 +183,6 @@ metadata:
 spec:
   template:
     spec:
-      version: v1.27.2-k0s.0
```
Thx. I did it.

kahirokunn commented on inttest/util/util.go:
```diff
@@ -437,3 +439,64 @@ func dowloadStableK0smotronOperator() (string, error) {
 }
+
+// GetK0smotronControlPlaneStatus retrieves the K0smotronControlPlane status
+func GetK0smotronControlPlaneStatus(ctx context.Context, client *kubernetes.Clientset, name, namespace string) (*cpv1beta1.K0smotronControlPlaneStatus, error) {
```
While working on many tasks, I left some unused code behind. I deleted it. Thank you 🙏

kahirokunn commented:
Thank you for your reviews and follow-ups😆
